git/index.go

```diff
@@ -199,7 +200,26 @@ func (g *Index) AddStage(c *Client, path IndexPath, mode EntryMode, s Sha1, stag
 			entry.Mtime = mtime
 			entry.Fsize = size
 
-			// We found and updated the entry, no need to continue
+			// If replace is true then we search for any entries that
+			//  should be replaced with this one. It could be that there
+			//  are entries with the same SHA-1 hash or there are children
+			//  that are orphaned because of the change from a directory to
+			//  a file.
+			if replaceEntry {
+				newObjects := make([]*IndexEntry, 0, len(g.Objects))
+				for _, e := range g.Objects {
```

Comment:
I'm sure this can be better than O(n^2), but for now it's better to get it working slowly than not have it at all.

Reply:
Right, perhaps we can do performance passes later on or as needed if in real uses it's too slow.